### ISSUE TITLE
"Killed" during node `build/ioq3ded.js` step

### ISSUE THREAD

**Issue by isaacbernat** [begleysm/quakejs#12]
I was following the steps is this Readme and the excellent [howto linked in the repo](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10) to install quakeJS into a raspberry pi zero 2 w. I got `setup_14.x` for n node.js instead of 8 and used `apt-get` instead of `apt` to have this running on Raspberry Pi OS (aka Rapsbian) so far.

Unfortunately it got the `Killed` message during step `node build/ioq3ded.js +set fs_game baseq3 +set dedicated 1`. The readme mentioned servers with less than 1 GB ram could die on that step. This machine has 512MB ram. Is there a way it could be installed? Maybe breaking that step into several smaller ones? Or copying the resulting files installed in another system directly here?

It would be really cool to spawn quakeJS "wifi-parties" on the spot. This card fits in one's wallet, as it's smaller than a credit card, so that would make it possible. This is why I wasn't trying a more powerful machine ^_^' (I tried the docker version on my Desktop computer and it worked fine, but I was afraid that would be too much overhead for the Raspberry and was trying to install it from scratch using these instructions here).

Thanks again for this nice repo!

**Comment by begleysm** [comment]
The [original quakejs](https://github.com/inolen/quakejs/) was developed by a guy named [Inolen](https://github.com/inolen/).  I just got it running again and tweaked it to operate entirely offline.

That being said, I'd never dug into the specifics of how ioq3ded.js works.  But you'll notice that that is the actually Quake 3 Server being run.  Node is the runtime environment, ioq3ded.js is the server software that it is running.  The rest are parameters being passed to ioq3ded.js.  So you're going to have to be able to run the server software to use your Raspberry Pi as a server.

It's possible that downloading the BASE3Q files is what is causing your system to crash.  Perhaps you could try pulling them from your Docker installation (or some other Linux installation on a beefier machine) and then pushing them to your Raspberry Pi device.  Maybe then you could run the command with the +exec server.cfg parameters and get it to go without downloading the BASE3Q files yourself.  If it still tries to download the files you may need to dig into ioq3ded.js and find where it checks to see if it needs to download the BASE3Q files.

**Comment by isaacbernat** [comment]
Copying the files from `/quakejs/base/baseq3` did the trick. Great! Thank you :D!

I have another issue now... looking at the logs, the server seems to be running fine. But when I open chrome to 192.168.1.144 (the local address assigned to the raspberry), I get the page with "quakeJS" logo and black background. When I check error console it mentions `Failed to load resource: net::ERR_NAME_NOT_RESOLVED quakejs/assets/manifest.json:1`. I also tried running the server on the pi adding `+set fs_cdn 192.168.1.144` (so the command becomes `node build/ioq3ded.js +set fs_game baseq3 +set dedicated 1 +exec server.cfg +set fs_cdn 192.168.1.144`) but no apparent changes. Manually going to http://192.168.1.144/assets/manifest.json delivers the file. I tried setting nano to "quakejs" also points to 127.0.0.1 (using the same pattern as content.quakejs.com mentioned in [linked howto](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10)) but again I didn't see any changes. I did follow the steps to `Setup a Local Content Server` and all other in the [linked howto](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10). Do you know what I could be missing?

**Comment by begleysm** [comment]
I have had questions in the past about it not working, everytime it has been user error of one type or another.  In fact just last week someone has the same problem and later figured out it was use error (although I don't know exactly what).

Are you sure you've followed the guide to the letter?  

There are 3 components, the play page, the content server, and the quakejs server.  You go to the play page in your browser, it downloads files from the content server, and then it connects to the quakejs server.

The page that you visited, and the JS logo page that you're seeing, is the play page.  Even if the quakejs server isn't running it should still ask you to accept an EULA, then download some files, then sit at a screen that says "CONNECTING TO <server address> AWAITING CHALLENGER .... #" with a red Quake 3 Logo in the background.

If this isn't happening, then you're not connecting to your content server properly.  Did you redirect [content.quakejs.com](https://content.quakejs.com) to your localhost as per here [https://steamforge.net/wiki/index.php/How\_to\_setup\_a\_local\_QuakeJS\_server\_under\_Debian\_9\_or\_Debian\_10#Redirect\_content.quakejs.com\_2](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10#Redirect_content.quakejs.com_2)?

Also, especially for "[The Simpler Method](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10#The_Simpler_Method)", you need to make sure your server is named "quakejs" on your local network.  If not you'll have to follow the instructions in "[The Original Method](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10#The_Original_Method)" and change some server reference names (basically everywhere it says "quakejs".

**Comment by isaacbernat** [comment]
**TLDR** it works now!

Thanks for the patience! I was following the "simpler method". I checked the docker implementation (which worked in my laptop before). I used these 2 lines from [the entrypoint.sh file](https://github.com/treyyoder/quakejs-docker/blob/master/entrypoint.sh) 
```
sed -i "s/quakejs/${SERVER}/g" index.html
/etc/init.d/apache2 start
```
replacing `${SERVER}` for 192.168.1.144 and restarting apache and everything seemed to work just fine. Did I miss some step on the tutorial? Or maybe it works differently on Raspbian? (e.g. like using `apt-get` instead of `apt get`).

> There are 3 components, the play page, the content server, and the quakejs server. You go to the play page in your browser, it downloads files from the content server, and then it connects to the quakejs server.

Afaik I always used the same URL, just the domain. So I guess this are redirects which happen transparent to the client? 

> The page that you visited, and the JS logo page that you're seeing, is the play page. Even if the quakejs server isn't running it should still ask you to accept an EULA, then download some files, then sit at a screen that says "CONNECTING TO AWAITING CHALLENGER .... #" with a red Quake 3 Logo in the background.

It didn't ask for the EULA until I did the "quakejs" sed change for the server "IP". I was trying to access this from another computer on the network, not from pi zero 2 (the poor machine can barely handle opening webpages and would die half-way installing/loading quakejs in browser).

> If this isn't happening, then you're not connecting to your content server properly. Did you redirect content.quakejs.com to your localhost as per here?

Yes, that part I did!

> you need to make sure your server is named "quakejs" on your local network

Could this have been my main problem? It is/was not very clear for me how to make this happen, so I thought following all the instructions there would be enough... Could you add instructions or a link where this is explained? I guess I am not too well versed in linux/networks and failed at that ^_^'

Thanks again!

**PS:** I think adding a line or two about troubleshooting in the "killed" section that copying the baseq3 files from somewhere might make this work on a 500MB ram machine like a pi zero 2 could be useful. Would you like me to amend the readme and create a PR? Or would you plane to add this info (and maybe on the tutorial too) yourself for future reference?

**Comment by begleysm** [comment]
Glad it's working.

I expanded the note in the README as you suggested.

As for your server name, yes that was probably your main issue.  You can check your hostname on a Debian-based machine (which Raspbian is) by typing `hostname` at the command line.  Here is an [article ](https://phoenixnap.com/kb/debian-change-hostname) that describes how to change your hostname.

The issue is that, to follow the guide, you need to be able to address your server as "quakejs".  In other words, you need to be able to "ping" quakejs and be able to go to http://quakejs to access your machine.  If you can access the webserver at http://192.168.1.144 but not http://quakejs then your hostname isn't set to quakejs.

You could have solved it by changing the machines hostname.  Instead you solved it by changing the references in the Play Page (index.html) from "quakejs" to "192.168.1.144" which is, essentially, what is described [Changing the Hostname](https://steamforge.net/wiki/index.php/How_to_setup_a_local_QuakeJS_server_under_Debian_9_or_Debian_10#Changing_the_Hostname) section of the guide (except you used an IP address instead of a different hostname).

**Comment by isaacbernat** [comment]
Thanks for the article! It all works perfect now using quakejs instead of the IP, which means I can use this with DHCP instead of a static IP. That's nicer :D

*PS:* I haven't tried but during my time debugging this I found this repo which claims is based in your instructions and automatised the process https://github.com/digidigital/quakejs-installer . Maybe I'll get another raspberry to try from scratch :D (I expect it will crash when downloading the files and I'll have to take that step manually, but now I know. Thanks again)

**Comment by begleysm** [comment]
Great! And thanks for the link to that installer project.  I wasn't aware of that but will link to it in my tutorial and README.
